Migrated Gaia DeviceStateTracker to seperate seng module

jesse-mm opened the pull request:
Would be nice to run some test using Travis. Coverage for this branch is 100%. Would be great to have some opinions about how it's setup. I didn't test the README steps but it should work as advertised testing this would require a NPM release (I think).

This version is intended to work with JavaScript as well as with TypeScript. Removed the IE9 deprecated stuff.

@flut1 Do you see any chance to update your project (seng-event) to the newest boilerplate version so typings is unnecessary, also I would like to know if it is possible to dispatch an event with payload, couldn't really find a method to do that.

@ThaNarie I've updated lint settings I'll make a PR for this as soon as possible the current config i've used in this project is a bit more forgiving.

jesse-mm committed:
Add unmodified DeviceStateTracker from Gaia Framework

jesse-mm committed:
Add seng-event as dependency

jesse-mm committed:
Remove matchMedia support check

jesse-mm committed:
Remove observables, minor refactoring

jesse-mm committed:
Allow variables to start with a _

jesse-mm committed:
Add interfaces for DeviceState and MediaQueries objects

jesse-mm committed:
Set new local variables add object types

jesse-mm committed:
Update class documentation

jesse-mm committed:
Only one line at EOF

jesse-mm committed:
Add device-state-tracker description

jesse-mm committed:
Add DeviceStateEvent

jesse-mm committed:
Add showStateIndicator option, dispatch DeviceStateEvent when changing DeviceState

jesse-mm committed:
Add configMock remove unused files

jesse-mm committed:
Add enum check change DeviceState interface

jesse-mm committed:
Add IMediaQuery interface in seperate file

jesse-mm committed:
Update imports

jesse-mm committed:
Update enum check

jesse-mm committed:
Update tests

jesse-mm committed:
Add additional test

jesse-mm committed:
Update tests

jesse-mm committed:
Update tests

jesse-mm committed:
Fix TS warning, object possible null

jesse-mm committed:
Comment why (native) matchMedia is replaced

jesse-mm committed:
Allows JS import for testing

Transforms JavaScript in test folder to ES2015 compatible source to run
in PhantomJS which only supports ES2015 syntax.

jesse-mm committed:
Update DeviceStateTracker to handle configMock in JavaScript (es6) format

jesse-mm committed:
Update docs

jesse-mm committed:
Add bundle

ThaNarie commented on README.md:
```diff
@@ -61,15 +60,92 @@ download as well:
+import { DeviceStateTracker, DeviceStateEvent } from 'seng-device-state-tracker';
+import { mediaQueries, DeviceState } from './path/to/config/deviceStateConfig';
+
+const deviceStateTracker:DeviceStateTracker = new DeviceStateTracker(mediaQueries, DeviceState);
```
`:DeviceStateTracker` should be removed in the JS example

ThaNarie commented on config/webpack.config.test.js:
```diff
@@ -34,7 +34,12 @@ module.exports = function()
-				}
+				},
+				{
+					test: /\.js$/,
```
Why do we need this, are we processing JS?

jesse-mm replied:
Yes for the test it's possible to import the JavaScript file with MediaQueries en DeviceState. Currently I commented out the JavaScript import from the test file. I did this because I actually didn't really know how to solve this. As the same code is tested for JavaScript I didn't want to create a whole new test file with the same testing methods only for testing the import. Any suggestions ?

ThaNarie commented on test/DeviceStateTrackerSpec.ts:
```diff
@@ -0,0 +1,140 @@
+
+			deviceStateTracker.addEventListener(DeviceStateEvent.STATE_UPDATE, eventHandler);
+
+			(<any> deviceStateTracker)._queryLists.forEach(mq => mq.callListeners());
```
How does this work?
Seems you are adding 1 event listener. Then looping trough the queries calling listener. Then the eventHandler removes the listener. Isn't the eventHandler called multiple times, for each callListener? How would that work?

jesse-mm replied:
We are testing a specific .matchMedia case here as I want to have to result for the MediaQuery defined by the test. The mock matchMedia doesn't allow some kind of resizing magic. I have to call the callListeners manually. Suggestions are welcome! :)

ThaNarie commented on test/DeviceStateTrackerSpec.ts:
```diff
@@ -0,0 +1,140 @@
+
+	it('should reverse the deviceState order', () => {
+		// Branch coverage
+		new DeviceStateTracker(mediaQueries, DeviceState, true);
```
So this doesn't test anything? Couldn't it test if the reverse order also works?

jesse-mm replied:
Nope doesn't test anything. I was thinking to reverse the array instead of looping using -1 but _queryList is still private so I'm not completly sure how to test it properly. 

What I could do though is making a second MediaQueryList and DeviceState that uses max-width to test it.

ThaNarie commented on test/DeviceStateTrackerSpec.ts:
```diff
@@ -0,0 +1,140 @@
+			},
+			[]);
+
+	(<any> deviceStateTracker)._queryLists.forEach((mq, index) => mq.media = mediaQueries[enumValues[index]]);
```
Are we accessing private variables here? Isn't that an test-anti-pattern?

jesse-mm replied:
Yes anti-pattern, but for now the mock MediaQueryList doesn't have a media key on it's list object so I need to append this for testing. I've opened a issue on the authors github but didn't have a reply yet.

ThaNarie commented on tslint.json:
```diff
@@ -5,7 +5,7 @@
     "no-console": 1,
     "prefer-array-literal": [ true, { "allow-type-parameters": true } ],
-    "variable-name": [ true, "allow-pascal-case" ],
+    "variable-name": [ true, "allow-pascal-case", "allow-leading-underscore" ],
```
Shouldn't we just ban the underscore for private vars?
In JS we don't ever do that anymore (airbnb doens't allow it either) and in TS we can just trust the private keyword.

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	/**
+	 * Array containing a boolean for each device state that indicates if the
+	 * media query currently matches. When multiple media queries match, we will
+	 * set the state to the one with the largest index.
```
> we will set the state to the one with the largest index.

Shouldn't that be the other way around when we reverse the order?

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	constructor(
+		mediaQueries:IMediaQuery,
+		deviceState:any,
+		reverseDeviceStateOrder:boolean = false,
```
Should we use an options object for these ones?

ThaNarie replied:
This is not done?

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+
+	/**
+	 * Checks if the enum object is correctly formatted. The DeviceState enum should only contain numeric values in
+	 * a ascending order.
```
a > an

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	 * matching breakpoint. This is usually convenient for mobile-first designs. If you want to reverse
+	 * this order (for desktop-first designs, for example). Pass the reverseDeviceStateOrder boolean as true.
+	 */
+	private _deviceState:any;
```
Shouldn't this at least be typed to `{ [key string]: any; }` ?

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+		// Get all the keys of deviceState object
+		const enumValues = Object.keys(deviceState).map(k => deviceState[k]);
+		// Get all the numeric values from enumValues
+		const enumKeys = enumValues.filter(value => typeof value === 'number') as number[];
```
`number[]` should be `Array<number>`

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+			throw new Error(`[DeviceStateTracker] DeviceState object should contain valid enum values`);
+		}
+
+		// Check if enum keys are in a ascending order
```
a > an

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	 * @param deviceState
+	 */
+	private enumCheck(deviceState:any):void {
+		let index = 0;
```
I would move this to the place it's actually used

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+		}
+
+		// Check if enum keys are in a ascending order
+		enumKeys.forEach((key:number) => {
```
why not use the second argument of the forEach as index? :)

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	 */
+	private initTracking():void {
+		this._deviceStateNames = Object.keys(this._deviceState).filter((key) => {
+			return isNaN(parseInt(key, 10));
```
unneeded function body with return

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+
+	/**
+	 * Loops through each deviceState and adds a matchMedia listener for each.
+	 * Calls updateFromMatchMedia_ to initialize the current state.
```
Why is that _ there?

jesse-mm replied:
No idea actually will remove it.

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	/**
+	 * Called whenever a MediaQueryList updates. Checks if the query matches
+	 * and stores the result in the queryListMatches_ array. Then calls
+	 * updateFromMatchMedia_() to find the current state from all matching
```
Why is that _ there?

jesse-mm replied:
No idea actually will remove it.

ThaNarie commented on src/lib/DeviceStateTracker.ts:
```diff
@@ -0,0 +1,213 @@
+	 * :before pseudo-element.
+	 */
+	private initStateIndicator():void {
+		this._stateIndicator = document.createElement('div');
```
this one is never removed on destruct?

jesse-mm committed:
Remove leading underscore notation for private variables

jesse-mm committed:
Remove DeviceStateTracker type from JS in README

jesse-mm committed:
DeviceState enum and MediaQueries are now in a configuration object.

Fixed tests, was still calling _queryLists.

jesse-mm committed:
Fix typos and docs

jesse-mm committed:
Use forEach index instead of a seperate var, update deviceState typing

jesse-mm committed:
Update typing []number to Array<number>

jesse-mm committed:
Remove function body use shorthand version

jesse-mm committed:
Remove state indicator from document when DeviceStateTracker destructs

jesse-mm committed:
Fix spelling mistake

jesse-mm committed:
Update docs

jesse-mm committed:
Update config mock with maxWidthQueries

Update test with max-width mediaQueries to test deviceStateTracker. Also
add a test for testing removal/non-removal of device state indicator.

jesse-mm committed:
Update readme

jesse-mm committed:
Update documentation

jesse-mm committed:
Update documentation

jesse-mm commented:
Bump @ThaNarie

jesse-mm committed:
Add DataEvent to sent device state data with event

jesse-mm committed:
Add test for DataEvent

jesse-mm committed:
Change public accessible data to private

jesse-mm committed:
update DeviceStateEvent and DeviceStateTracker tests

jesse-mm committed:
change constructor to only accept a configuration object

jesse-mm committed:
Update README

jesse-mm commented:
Update PR! :)

jesse-mm committed:
Fix readme formatting

jesse-mm committed:
Fix object typo in README

jesse-mm committed:
update readme

jesse-mm committed:
update travis configuration
